File: vision/branches/20090415/src/java/cast/server/CASTComponentFactory.java
package cast.server;

import Ice.Current;
import Ice.Identity;
import Ice.Object;
import Ice.ObjectAdapter;
import Ice.ObjectPrx;
import cast.ComponentCreationException;
import cast.core.CASTUtils;
import cast.interfaces.CASTComponentPrx;
import cast.interfaces.ManagedComponentPrx;
import cast.interfaces.ManagedComponentPrxHelper;
import cast.interfaces.TaskManagerPrx;
import cast.interfaces.TaskManagerPrxHelper;
import cast.interfaces.UnmanagedComponentPrx;
import cast.interfaces.UnmanagedComponentPrxHelper;
import cast.interfaces.WorkingMemoryPrx;
import cast.interfaces.WorkingMemoryPrxHelper;
import cast.interfaces._ComponentFactoryDisp;

/**
 * 
 * Server that creates components
 * 
 * @author nah
 * 
 */
public class CASTComponentFactory extends _ComponentFactoryDisp {

	public CASTComponentPrx newComponent(String _id, String _type,
			Current __current) throws ComponentCreationException {
		ObjectPrx base = createBase(_id, _type, __current);
		return WorkingMemoryPrxHelper.checkedCast(base);
	}

	public ManagedComponentPrx newManagedComponent(String _id, String _type,
			Current __current) throws ComponentCreationException {
		ObjectPrx base = createBase(_id, _type, __current);
		return ManagedComponentPrxHelper.checkedCast(base);
	}

	public TaskManagerPrx newTaskManager(String _id, String _type,
			Current __current) throws ComponentCreationException {
		ObjectPrx base = createBase(_id, _type, __current);
		return TaskManagerPrxHelper.checkedCast(base);
	}

	public UnmanagedComponentPrx newUnmanagedComponent(String _id,
			String _type, Current __current) throws ComponentCreationException {
		ObjectPrx base = createBase(_id, _type, __current);
		return UnmanagedComponentPrxHelper.checkedCast(base);
	}

	public WorkingMemoryPrx newWorkingMemory(String _id, String _type,
			Current __current) throws ComponentCreationException {
		ObjectPrx base = createBase(_id, _type, __current);
		return WorkingMemoryPrxHelper.checkedCast(base);
	}

	private ObjectPrx createBase(String _id, String _type, Current __current)
			throws ComponentCreationException {
		try {
			ObjectAdapter adapter = __current.adapter;
			Identity id = new Identity(_id, _type);
			Object servant = CASTUtils.createServant(id, adapter);
			ObjectPrx base = adapter.add(servant, id);
			return base;
		} catch (ClassNotFoundException e) {
			ComponentCreationException cce = new ComponentCreationException(
					"Unable to create component: Cannot find class "
							+ e.getMessage());
			cce.initCause(e);
			throw cce;
		} catch (Exception e) {
			ComponentCreationException cce = new ComponentCreationException(
					"Unable to create component: " + e.getClass() + ": "
							+ e.getMessage());
			cce.initCause(e);
			throw cce;
		}
	}

}

File: comsys/tags/release1/tools/cast/src/c++/cast/server/CASTComponentFactory.cpp
#include "CASTComponentFactory.hpp"

#include <CDL.hpp>
#include <ComponentCreator.hpp>
#include <iostream>


using namespace Ice;
using namespace cast::cdl;
using namespace cast::interfaces;
using namespace std;

namespace cast {

  
  cast::interfaces::CASTComponentPrx 
  CASTComponentFactory::createBase(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt) 
    throw (ComponentCreationException)
  {

    try {
      //get creator
      DynamicComponentCreator * dcc(NULL);
      ComponentCreatorMap::iterator i = m_creators.find(type);
      //if we don't have it yet
      if(i == m_creators.end()) {
	dcc = createComponentCreator(type);
	m_creators[type] = dcc;      
      }
      else {
	dcc =  i->second;
      }
      
      Identity iceid;
      iceid.name = id;
      iceid.category = type;    
      
      cast::CASTComponentPtr component = dcc->createNewComponent(id);
      component->_setObjectAdapter(_crt.adapter);
      component->_setIceIdentity(iceid);
      component->_setComponentPointer(component);
      
      ObjectPrx base = _crt.adapter->add(component,iceid);
      
      return CASTComponentPrx::uncheckedCast(base);
    }
    catch (const CASTException & e) {
      cerr<<"Exception when creating component "<<id<<" of type "<<type<<endl;
      cerr<<e.message<<endl;
      ComponentCreationException cce(e.message);
      throw cce;
    }    
  }
  
  cast::interfaces::CASTComponentPrx CASTComponentFactory::newComponent(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt) 
    throw (ComponentCreationException) {
    return createBase(id,type,_crt);
  }
  
  cast::interfaces::ManagedComponentPrx CASTComponentFactory::newManagedComponent(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt) 
    throw (ComponentCreationException) {
    return ManagedComponentPrx::checkedCast(createBase(id,type,_crt));
  }
  
  cast::interfaces::UnmanagedComponentPrx CASTComponentFactory::newUnmanagedComponent(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt) 
    throw (ComponentCreationException) {
    return UnmanagedComponentPrx::checkedCast(createBase(id,type,_crt));    
  }
  
  cast::interfaces::WorkingMemoryPrx CASTComponentFactory::newWorkingMemory(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt) 
    throw (ComponentCreationException) {
    return WorkingMemoryPrx::checkedCast(createBase(id,type,_crt));
  }
  
  cast::interfaces::TaskManagerPrx CASTComponentFactory::newTaskManager(const ::std::string& id, const ::std::string& type, const ::Ice::Current & _crt)     
    throw (ComponentCreationException) {
    return TaskManagerPrx::checkedCast(createBase(id,type,_crt));
  }


//   void 
//   CASTComponentFactory::remove(const std::string & _id, 
// 				const std::string & _type,
// 				const Ice::Current & _crt) {
//     printf("CASTComponentFactory::destroy(): %s %s\n", _id.c_str(), _type.c_str());
//     Identity id;
//     id.name = _id;
//     id.category = _type;
//     _crt.adapter->remove(id);
//   }

  
};
